### PR TITLE
Unique names for data classes and sample types

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
For this [story ](https://docs.google.com/document/d/1N6xfULI198pbU7xYdijYi4jmEg8n1a9s5PcPc5Tb4w0/edit) we want to support an administrative ability to set an optional prefix for data classes and sample types on a per container basis. 

We also will want to give administrators the ability to disable manually inserting/updating a name for sample types and data classes. This should make is so the name field is removed from all insert/update forms for LKS, LKB, LKSM and raise an error if a name comes in through an API endpoint.

#### Related Pull Requests
- https://github.com/LabKey/sampleManagement/pull/723
- https://github.com/LabKey/labkey-ui-components/pull/652
- https://github.com/LabKey/biologics/pull/1023

#### Changes
- Define a `NameExpressionOptionService` with a default NoOp implementation. An actual implementation will be provided in the `SampleManagementModule`
- If users are not allowed to manually insert names
  - Remove the name field for insert/update views
  - Add validation in the name expression `DataIterator` to fail if detected
  - Disallow removing the name expression from the domain editors
  - For samples, if the `UpdateExisting` option is specified, fail only if a specified name doesn't exist in the database

### CODE REVIEW

**Comment by cnathe** [comment]
@labkey-klum I'm not sure if it was part of this story or not, but Rosaline and I were talking about some changes in the Field editor related to this container prefix and the per sample type naming pattern. She was demoing the container prefix part to me and mentioned that if the sample type had a naming pattern and then the container prefix was updated, then that updated prefix would override and change the per sample type naming pattern. Is this as expected? For example, if my sample type had a naming pattern of "PREFIX-Blood-${genId}" and the container prefix was changed to "PRX", I would not have expected the sample type naming pattern to then be set to "PRX". I would see it being updated to "PRX-Blood-${genId}". Just wanted to check in on this. Thanks.

File: api/src/org/labkey/api/dataiterator/NameExpressionDataIterator.java [hunk 1]
```diff
@@ -93,7 +100,11 @@ public Object get(int i)
                 curName = StringUtils.isEmpty((String)curName) ? null : curName;
 
             if (curName != null)
+            {
+                if (!_allowUserSpecifiedNames)
+                    getErrors().addRowError(new ValidationException("Manual entry of names has been disabled for this folder. Only naming pattern generated names are allowed."));
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/dataiterator/NameExpressionDataIterator.java]:
```suggestion
                    getErrors().addRowError(new ValidationException("Manual entry of names has been disabled for this folder. Only naming-pattern-generated names are allowed."));
```
Or, if you're anti-hyphen, "Only names generated by naming patterns are allowed."

File: api/src/org/labkey/api/exp/api/NameExpressionOptionService.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    String getExpressionPrefix(Container c);
+
+    /**
+     * Set the name expression prefix for the specified container. Setting the prefix to null
+     * will clear out an existing prefix.
+     */
+    void setExpressionPrefix(Container c, User user, @Nullable String prefix) throws Exception;
+
+    /**
+     * Returns whether user specified names for samples and dataclasses are permitted for this folder.
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/exp/api/NameExpressionOptionService.java]:
```suggestion
     * Returns whether user-specified names for samples and dataclasses are permitted for this folder.
```

File: api/src/org/labkey/api/exp/api/NameExpressionOptionService.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    boolean allowUserSpecifiedNames(Container c);
+
+    /**
+     * Set whether user specified names for samples and dataclasses are allowed for this folder. If set to false
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/exp/api/NameExpressionOptionService.java]:
```suggestion
     * Set whether user-specified names for samples and dataclasses are allowed for this folder. If set to false
```

File: experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java [hunk 4]
```diff
@@ -205,6 +206,13 @@ public MutableColumnInfo createColumn(String alias, Column column)
                 c.setNullable(nameExpression != null);
                 String desc = ExpMaterialTableImpl.appendNameExpressionDescription(c.getDescription(), nameExpression);
                 c.setDescription(desc);
+
+                // shut off this field in insert and update views if user specified names are not allowed
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java]:
```suggestion
                // shut off this field in insert and update views if user-specified names are not allowed
```

File: api/src/org/labkey/api/exp/api/NameExpressionOptionService.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+        @Override
+        public void setAllowUserSpecifiedNames(Container c, User user, boolean allowNames) throws Exception
+        {
+        }
+
+        @Override
+        public String createPrefixedExpression(Container c, String nameExpression, boolean isAliquotNameExpression)
+        {
+            return null;
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/exp/api/NameExpressionOptionService.java]: Should this return `nameExpression`?

File: experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java [hunk 6]
```diff
@@ -856,6 +869,22 @@ else if (aliquotedFromObj instanceof Number)
...
+                Object currNameObj = map.get("Name");
+                if (currNameObj != null && !_allowUserSpecifiedNames)
+                {
+                    if (StringUtils.isNotBlank(currNameObj.toString()))
+                    {
+                        if (_context.getInsertOption().equals(QueryUpdateService.InsertOption.MERGE))
+                        {
+                            // don't flag rows that already exist if the option is set to update existing
+                            if (!rowExists(currNameObj.toString()))
+                                addRowError("Manual entry of names has been disabled for this folder. Only naming pattern generated names (or existing names) are allowed.");
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java]:
```suggestion
                                addRowError("Manual entry of names has been disabled for this folder. Only naming-pattern- generated names (or existing names) are allowed.");

```

File: experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java [hunk 7]
```diff
@@ -856,6 +869,22 @@ else if (aliquotedFromObj instanceof Number)
...
+                        }
+                        else
+                            addRowError("Manual entry of names has been disabled for this folder. Only naming pattern generated names are allowed.");
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java]:
```suggestion
                            addRowError("Manual entry of names has been disabled for this folder. Only naming-pattern-generated names are allowed.");

```

File: experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java [hunk 8]
```diff
@@ -856,6 +869,22 @@ else if (aliquotedFromObj instanceof Number)
...
 
+                Object currNameObj = map.get("Name");
+                if (currNameObj != null && !_allowUserSpecifiedNames)
+                {
+                    if (StringUtils.isNotBlank(currNameObj.toString()))
+                    {
+                        if (_context.getInsertOption().equals(QueryUpdateService.InsertOption.MERGE))
+                        {
+                            // don't flag rows that already exist if the option is set to update existing
+                            if (!rowExists(currNameObj.toString()))
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/samples/UploadSamplesHelper.java:880]: This seems like the sort of thing the ExistingRecordDataIterator was designed for, but I don't think it's in the mix here. This stashed set of names is probably sufficient, even if there's an outside chance the name won't be in the selected set.

**Review by labkey-susanh — APPROVED** [review]

File: api/src/org/labkey/api/exp/api/NameExpressionOptionService.java (hunk 5 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/exp/api/NameExpressionOptionService.java]: yes good catch

File: experiment/src/org/labkey/experiment/api/SampleTypeServiceImpl.java [hunk 9]
```diff
@@ -627,6 +630,17 @@ public ExpSampleTypeImpl createSampleType(Container c, User u, String name, Stri
...
 
+        NameExpressionOptionService svc = NameExpressionOptionService.get();
+        if (!svc.allowUserSpecifiedNames(c))
+        {
+            if (nameExpression == null)
+                throw new ExperimentException(NAME_EXPRESSION_REQUIRED_MSG);
+
+            // automatically apply the configured prefix to the name expression
+            nameExpression = svc.createPrefixedExpression(c, nameExpression, false);
+            aliquotNameExpression = svc.createPrefixedExpression(c, aliquotNameExpression, true);
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/SampleTypeServiceImpl.java:644]: I'm not sure the applying of prefixes should happen only when we don't allow user-specified names.  It seems like these settings are meant to be independent of each other.

**Review by labkey-susanh — CHANGES_REQUESTED** [review]

File: experiment/src/org/labkey/experiment/api/SampleTypeServiceImpl.java (hunk 9 above)
Comment (labkey-klum) [reply experiment/src/org/labkey/experiment/api/SampleTypeServiceImpl.java:644]: This is true, I've split it out from that conditional

**Review by labkey-susanh — APPROVED** [review]

**Pull request merged**